### PR TITLE
KAFKA-13340: Change ZooKeeperTestHarness to QuorumTestHarness

### PR DESCRIPTION
**Pull request by cmccabe**
Change ZooKeeperTestHarness to QuorumTestHarness so that integration tests which inherit from this
class can test both ZK and KRaft mode. Test cases which do this can specify the modes they support
by including a ParameterizedTest annotation before each test case, like the following:

@ParameterizedTest
@valuesource(strings = Array("zk", "kraft"))
def testValidCreateTopicsRequests(quorum: String): Unit = { ... }

For each value that is specified here (zk, kraft), the test case will be run once in the appropriate
mode. So the test shown above is run twice. This allows integration tests to be incrementally
converted over to support KRaft mode, rather than rewritten to support it. As you might expect, test
cases which do not specify a quorum argument will continue to run only in ZK mode.

JUnit5 makes the quorum annotation visible in the TestInfo object which each @beforeeach function in
a test can optionally take. Therefore, this PR converts over the setUp function of the quorum base
class, plus every derived class, to take a TestInfo argument. The TestInfo object gets "passed up
the chain" to the base class, where it determines which quorum type we create (zk or kraft).

The general approach taken here is to make as much as possible work with KRaft, but to leave some
things as ZK-only when appropriate. For example, a test that explicitly requests an AdminZkClient
object will get an exception if it is running in KRaft mode. Similarly, tests which explicitly
request KafkaServer rather than KafkaBroker will get an exception when running in KRaft mode.

As a proof of concept, this PR converts over MetricsTest to support KRaft.

### CODE REVIEW

File: core/src/test/scala/kafka/utils/TestInfoUtils.scala [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+      true
+    } else if (testInfo.getDisplayName().contains("quorum=zk")) {
+      false
+    } else {
+      throw new RuntimeException(s"Unknown quorum value")
+    }
+  } else {
+    false
+  }
+}
```
Comment (soarez) [inline core/src/test/scala/kafka/utils/TestInfoUtils.scala]: Short indent

```suggestion
  def isKRaft(testInfo: TestInfo): Boolean = {
    if (testInfo.getDisplayName().contains("quorum=")) {
      if (testInfo.getDisplayName().contains("quorum=kraft")) {
        true
      } else if (testInfo.getDisplayName().contains("quorum=zk")) {
        false
      } else {
        throw new RuntimeException(s"Unknown quorum value")
      }
    } else {
      false
    }
  }
```

File: core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala [hunk 2]
```diff
@@ -0,0 +1,339 @@
...
+
+  def zkClientOrNull: KafkaZkClient = if (isKRaftTest()) null else asZk().zkClient
+
+  def adminZkClient: AdminZkClient = asZk().adminZkClient
+
+  def zkPort: Int = asZk().zookeeper.port
+
+  def zkConnect: String = s"127.0.0.1:$zkPort"
+
+  def zkConnectOrNull: String = if (isKRaftTest()) null else s"127.0.0.1:$zkPort"
```
Comment (soarez) [inline core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala]:
```suggestion
  def zkConnectOrNull: String = if (isKRaftTest()) null else zkConnect
```

File: core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala [hunk 3]
```diff
@@ -0,0 +1,339 @@
...
+  def shutdownZooKeeper(): Unit = asZk().shutdownZooKeeper()
+
+  private def formatDirectories(directories: immutable.Seq[String],
+                                metaProperties: MetaProperties): Unit = {
+    val stream = new ByteArrayOutputStream()
+    var out: PrintStream = null
+    try {
+      out = new PrintStream(stream)
+      if (StorageTool.formatCommand(out, directories, metaProperties, false) != 0) {
+        throw new RuntimeException(out.toString())
```
Comment (soarez) [inline core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala]: Wrong variable

```suggestion
        throw new RuntimeException(stream.toString())
```

File: core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala [hunk 4]
```diff
@@ -0,0 +1,339 @@
...
+        metrics = controllerMetrics,
+        threadNamePrefix = Option(threadNamePrefix),
+        controllerQuorumVotersFuture = controllerQuorumVotersFuture)
+      controllerServer.socketServerFirstBoundPortFuture.whenComplete((port, e) => {
+        if (e != null) {
+          error("Error completing controller socket server future", e)
+          controllerQuorumVotersFuture.completeExceptionally(e)
+        } else {
+          controllerQuorumVotersFuture.complete(Collections.singletonMap(1000,
+            new InetAddressSpec(new InetSocketAddress("localhost", port))));
```
Comment (hachikuji) [inline core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala]: nit: unneeded semicolons here and below

**Review by hachikuji — COMMENTED** [review]
The patch seems ok. I confess it's not obvious to me what this approach buys us over the use of `ClusterTestExtensions`. Seems like another way we could do this is to let `QuorumTestHarness` make use of `ClusterTestExtensions` instead of duplicating all the logic to build the cluster. That would still let us make minor modifications to the extensions in order to enable it. As it is, we're left with two separate approaches for covering kraft and zk which seems less than ideal. It might be fine, but it would be nice to understand what was the gap with `ClusterTestExtensions` and whether we should fix it.

File: core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala (hunk 4 above)
Comment (cmccabe) [reply core/src/test/scala/integration/kafka/server/QuorumTestHarness.scala]: fixed

**Comment by cmccabe** [comment]
> The patch seems ok. I confess it's not obvious to me what this approach buys us over the use of ClusterTestExtensions. Seems like another way we could do this is to let QuorumTestHarness make use of ClusterTestExtensions instead of duplicating all the logic to build the cluster. That would still let us make minor modifications to the extensions in order to enable it. As it is, we're left with two separate approaches for covering kraft and zk which seems less than ideal. It might be fine, but it would be nice to understand what was the gap with ClusterTestExtensions and whether we should fix it.

I did consider reusing one of the other test frameworks, but there just isn't a lot to be gained because of the radically different approach that existing tests take. The code for creating the `BrokerServer` / `ControllerServer` is actually a VERY tiny part of this PR... the majority of the PR is integrating with stuff like the `@BeforeEach` functions in each test, handling the way the control flow bounces around the inheritance hierarchy, etc.

The goal here is to be able to convert existing tests without rewriting them. I have actually done this for about a dozen tests, which I haven't posted here since this PR is big enough. I think this is the only approach that will get us over the finish line for kraft. The effort to move to a saner test structure, not based on N levels of inheritance (which partially duplicate each others' functionality), is a great goal but I think we should pursue it for NEW tests rather than existing.

**Review by hachikuji — APPROVED** [review]
LGTM assuming tests are passing.

**Comment by cmccabe** [comment]
I had to fix a bug where ServerShutdownTest was hanging because we didn't call `zkClient#close`. This was causing Jenkins test runs to go off the rails. Hopefully the next test run will be good.

**Pull request merged**